Complete pose publishing

cosmicog opened the pull request:
Hello again, 

- There was a topic advertised but there was no publishing. AFAICS, your lib doesn't expose covariance matrix, so I left it blank.
 
- ~~Added `loc2d.launch` file with default params.~~

- ~~Also removed `/` in the prefixes of global namespace topics because my robot has robot namespaced topics and I had problems remapping them.~~

eupedrosa commented:
Hi @cosmicog.

Your PR does more than *Complete pose publishing*. I woul like to see this PR only focused on adding the *pose publishing*.

> There was a topic advertised but there was no publishing. AFAICS, your lib doesn't expose covariance matrix, so I left it blank.

You are correct, the lib does not exposes the Covariance matrix, at least for now -- quick inside info, it will be exposed.

> * Added loc2d.launch file with default params.
> * Also removed / in the prefixes of global namespace topics because my robot has robot namespaced topics and I had problems remapping them.

Please provide a different PR for each topic. This way we can discuss each individual topic.

cosmicog commented:
Ok. done. Also note: I made the topic latched in order to inform other nodes about latest published pose regardless of time of the subscription.
```cpp
pose_pub_ = nh_.advertise<geometry_msgs::PoseWithCovarianceStamped>("/pose", 2, true);
```

cosmicog commented:
> quick inside info, it will be exposed.

Nice. When it will arrive, you can add it to the `cur_pose_msg_`.

cosmicog commented:
I have forgotten adding the frame_id and time stamp 2 days ago. Added them, also moved pose publishing into the update section. I checked now and amcl does the same thing, doesn't publish to pose topic when there is no update. I thought mirroring its behaviour is the best but what's your opinion?

eupedrosa commented on include/lama/ros/loc2d_ros.h:
```diff
@@ -141,6 +143,9 @@ class Loc2DROS {
     Pose2D initial_prior_;
+    Pose2D current_pose_;
+    tf::Quaternion current_orientation_;
+    geometry_msgs::PoseWithCovarianceStamped cur_pose_msg_;
```
We do not need to keep track of these 3 variables. The Pose of the robot should be published only after its calculation.

This is a semantic issue. The `tf` will have a "continuous" version of the robot's pose, while the topic `~pose` should contain only the pose calculated by the Localization process.

cosmicog replied:
- **`geometry_msgs::PoseWithCovarianceStamped cur_pose_msg_;`**
  In general I like to keep memory allocated, so I allocate everything in constructors. With the correct compiler flags being set, even we allocate memory for an object in a function, it is allocated and reused after those function calls. But this time I did it for another reason: to avoid setting; 
  ```cpp
  cur_pose_msg_.header.frame_id = global_frame_id_;
  ```
  everytime we publish this message, I made it a member since this will never change.

- **`Pose2D current_pose_;`**
  You are right about having it since getPose() returns a reference, sorry my bad. I removed it.

- **`tf::Quaternion current_orientation_;`***
  But about this quaternion I'd like to keep it and avoid calling `tf::createQuaternionFromYaw(double yaw)` and calculating the quaternion twice in every update:
  - first in `onLaserScan()`
  - second in `publishCurrentPose()`
  
- **The pose topic logic**
  > This is a semantic issue. The tf will have a "continuous" version of the robot's pose, while the topic ~pose should contain only the pose calculated by the Localization process.

  I somehow agree with this semantic issue but I have a robot that travels between floors and changes maps. `amcl` follows the the same semantic. Which is quite ok for me since my robot's software stack requires the current belief information of the robot being in the `pose` topic. Even though robot didn't move and no updates have been computed, I still need the initial pose values in the `pose` topic. This way I can start navigating and knowing my pose in the beginning of the time that a new instances of move_base and map_server is started running. This way my software stack works with `amcl` and `loc_2d_ros` just fine.
  
  `tf` requires informations to have an updated timestamp, so even though no update, we still publish information to it. But this `pose` topic is for just informing other nodes about last known pose and doesn't require the updated timestamp information. `amcl` stops publishing to this topic when there is no update, also publishes once after every manual pose estimate message.

eupedrosa replied:
> Even though robot didn't move and no updates have been computed, I still need the initial pose values in the pose topic

Fair enough

eupedrosa commented on src/loc2d_ros.cpp:
```diff
@@ -83,6 +83,14 @@ lama::Loc2DROS::Loc2DROS()
+    cur_pose_msg_.pose.pose.orientation.z = tf::createQuaternionFromYaw(init_a).getZ();
+    cur_pose_msg_.pose.pose.orientation.w = tf::createQuaternionFromYaw(init_a).getW();
+    cur_pose_msg_.header.stamp = ros::Time::now();
+    pose_pub_.publish(cur_pose_msg_);
```
In my opinion we do not need to keep a MSG, we can create a new message each time we need to publish.
This block of code is not needed.

cosmicog replied:
I agree, I removed this since in the beginning, we are going to publish it inside `onInitialPose(const Pose2D& prior)`

eupedrosa commented on src/loc2d_ros.cpp:
```diff
@@ -229,11 +253,12 @@ void lama::Loc2DROS::onLaserScan(const sensor_msgs::LaserScanConstPtr& laser_sca
         tf::Stamped<tf::Pose> odom_to_map;
         try{
-            tf::Transform tmp_tf(tf::createQuaternionFromYaw(pose.rotation()), tf::Vector3(pose.x(), pose.y(), 0));
+            tf::Transform tmp_tf(current_orientation_, tf::Vector3(current_pose_.x(), current_pose_.y(), 0));
```
No need to do this.

cosmicog replied:
Updated.

eupedrosa commented on src/loc2d_ros.cpp:
```diff
@@ -140,6 +148,21 @@ lama::Loc2DROS::~Loc2DROS()
+    {
+        current_pose_ = loc2d_.getPose();
+        current_orientation_ = tf::createQuaternionFromYaw(current_pose_.rotation());
+    }
```
Why the need for the `init`? This function should always get the current pose from the `loc2d_` object.

cosmicog replied:
Right, with other updates I mentioned in comments, I deleted this.

eupedrosa commented on src/loc2d_ros.cpp:
```diff
@@ -140,6 +148,21 @@ lama::Loc2DROS::~Loc2DROS()
+    cur_pose_msg_.pose.pose.position.x = current_pose_.x();
+    cur_pose_msg_.pose.pose.position.y = current_pose_.y();
+    cur_pose_msg_.pose.pose.orientation.z = current_orientation_.getZ();
+    cur_pose_msg_.pose.pose.orientation.w = current_orientation_.getW();
```
This is not the correct way the update the orientation of the pose. It should be done with an appropriate function:
```c++
tf::quaternionTFToMsg(tf::createQuaternionFromYaw(current_pose.rotation()), cur_pose_msg.pose.pose.orientation);
```
This will create the propriate quaternion from a *yaw*.

cosmicog replied:
I kept the `current_orientation_` variable like I explained above to avoid computing the quaternion twice by calling `tf::createQuaternionFromYaw(double yaw)`

eupedrosa commented:
I think we have a go with this PR..

> Nice. When it will arrive, you can add it to the cur_pose_msg_.

I will.
